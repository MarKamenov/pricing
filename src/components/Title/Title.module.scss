.title {
    margin: 0;
    line-height: 1.2;
}

// Alignment variants
.align-left {
    text-align: left;
}

.align-center {
    text-align: center;
}

.align-right {
    text-align: right;
}

// Size variants
.size-sm {
    font-size: var(--font-size-sm);
}

.size-md {
    font-size: var(--font-size-md);
}

.size-lg {
    font-size: var(--font-size-lg);
}

.size-xl {
    font-size: var(--font-size-xl);
}

.size-2xl {
    font-size: var(--font-size-2xl);

    @media (min-width: 768px) {
        font-size: calc(var(--font-size-2xl) * 1.2);
    }
}

.size-3xl {
    font-size: var(--font-size-3xl);

    @media (min-width: 768px) {
        font-size: calc(var(--font-size-3xl) * 1.2);
    }
}

// Weight variants
.weight-normal {
    font-weight: 400;
}

.weight-medium {
    font-weight: 500;
}

.weight-semibold {
    font-weight: 600;
}

.weight-bold {
    font-weight: 700;
}

// Color variants
.color-primary {
    color: var(--text-page-title);
}

.color-secondary {
    color: var(--text-secondary);
}

.color-accent {
    color: var(--accent);
}

.color-default {
    color: var(--text-primary);
}

// Highlight styles
.highlight {
    color: var(--text-page-title-part);
    display: inline-block;
}

.subtitle {
    text-align: center;
    color: var(--text-page-subtitle)
}