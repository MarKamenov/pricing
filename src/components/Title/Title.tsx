import type { ReactNode } from "react"
import titleStyles from "./Title.module.scss"

interface TitleProps {
    children: ReactNode
    highlight?: string | string[]
    subtitle?: string
    align?: "left" | "center" | "right"
    size?: "sm" | "md" | "lg" | "xl" | "2xl" | "3xl"
    weight?: "normal" | "medium" | "semibold" | "bold"
    color?: "primary" | "secondary" | "accent" | "default"
    className?: string
    as?: "h1" | "h2" | "h3" | "h4" | "h5" | "h6"
}

export const Title = ({
    children,
    highlight,
    align = "left",
    size = "xl",
    weight = "bold",
    color = "primary",
    className = "",
    subtitle,
    as: Component = "h1",
}: TitleProps) => {
    const titleClasses = `
    ${titleStyles.title}
    ${titleStyles[`align-${align}`]}
    ${titleStyles[`size-${size}`]}
    ${titleStyles[`weight-${weight}`]}
    ${titleStyles[`color-${color}`]}
    ${className}
  `.trim()

    if (!highlight) {
        return <Component className={titleClasses}>{children}</Component>
    }

    const highlights = Array.isArray(highlight) ? highlight : [highlight]
    const text = children?.toString() ?? ""

    // Split text by highlights and create an array of regular text and highlighted spans
    const parts = text.split(new RegExp(`(${highlights.join("|")})`, "gi"))

    return (
        <div>
            <Component className={titleClasses}>
                {parts.map((part, index) => {
                    const isHighlighted = highlights.some((h) => part.toLowerCase() === h.toLowerCase())

                    return isHighlighted ? (
                        <span key={index} className={titleStyles.highlight}>
                            {part}
                        </span>
                    ) : (
                        part
                    )
                })}

            </Component>
            {subtitle && <p className={titleStyles.subtitle}>{subtitle}</p>}
        </div>

    )
}

