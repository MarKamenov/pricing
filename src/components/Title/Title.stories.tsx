import { Meta, StoryFn } from '@storybook/react';
import { Title } from './Title';

const meta: Meta<typeof Title> = {
    title: 'Components/Title',
    component: Title,
    argTypes: {
        children: { control: 'text' },
        highlight: { control: 'text' },
        subtitle: { control: 'text' },
        align: { control: { type: 'select', options: ['left', 'center', 'right'] } },
        size: { control: { type: 'select', options: ['sm', 'md', 'lg', 'xl', '2xl', '3xl'] } },
        weight: { control: { type: 'select', options: ['normal', 'medium', 'semibold', 'bold'] } },
        color: { control: { type: 'select', options: ['primary', 'secondary', 'accent', 'default'] } },
        className: { control: 'text' },
        as: {
            control: { type: 'select', options: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'] },
        },
    },
};

export default meta;

type Story = StoryFn<typeof Title>;

// Primary Story with the rendering implementation
export const Primary: Story = (args) => <Title {...args} />;

Primary.args = {
    children: 'This is a Primary Title',
    highlight: 'Primary',
    subtitle: 'This is a subtitle for the primary title',
    align: "left",
    size: "xl",
    weight: "bold",
    color: "primary",
    className: "",
    as: "h1",
};

// Secondary Story with the rendering implementation
export const Secondary: Story = (args) => <Title {...args} />;

Secondary.args = {
    children: 'Secondary Title Example',
    highlight: 'Secondary',
    subtitle: 'A subtitle to support the secondary title.',
    color: 'secondary',
    align: "left",
    size: "xl",
    weight: "bold",
    className: "",
    as: "h1",
};

// Accent Story with the rendering implementation
export const Accent: Story = (args) => <Title {...args} />;
Accent.args = {
    children: 'Accent Title Here',
    highlight: 'Accent',
    subtitle: 'More details about accent.',
    color: 'accent',
    align: "left",
    size: "xl",
    weight: "bold",
    className: "",
    as: "h1",
};

// Default Story with the rendering implementation
export const Default: Story = (args) => <Title {...args} />;
Default.args = {
    children: 'Default Title Style',
    highlight: 'Title',
    subtitle: 'Subtitle for the default style.',
    color: 'default',
    align: "left",
    size: "xl",
    weight: "bold",
    className: "",
    as: "h1",
};

// Custom Size and Weight Story with the rendering implementation
export const CustomSizeAndWeight: Story = (args) => <Title {...args} />;
CustomSizeAndWeight.args = {
    children: 'Custom Sized Title',
    highlight: 'Sized',
    subtitle: 'Showing off custom sizes.',
    size: '3xl',
    weight: 'normal',
    color: "primary",
    align: "left",
    className: "",
    as: "h1",
};

// Centered Title Story with the rendering implementation
export const CenteredTitle: Story = (args) => <Title {...args} />;
CenteredTitle.args = {
    children: 'Centered Title',
    highlight: 'Centered',
    subtitle: 'This title is aligned to the center.',
    align: 'center',
    size: "xl",
    weight: "bold",
    color: "primary",
    className: "",
    as: "h1",
};

// Multiple Highlights Story with the rendering implementation
export const MultipleHighlights: Story = (args) => <Title {...args} />;
MultipleHighlights.args = {
    children: 'Multiple Highlights Title',
    highlight: ['Multiple', 'Highlights'],
    subtitle: 'Shows multiple highlighted words.',
    align: "left",
    size: "xl",
    weight: "bold",
    color: "primary",
    className: "",
    as: "h1",
};

// As H2 Story with the rendering implementation
export const AsH2: Story = (args) => <Title {...args} />;
AsH2.args = {
    children: 'Title rendered as H2',
    highlight: 'H2',
    subtitle: 'Demonstrates the "as" prop.',
    as: 'h2',
    align: "left",
    size: "xl",
    weight: "bold",
    color: "primary",
    className: "",
};

// With ClassName Story with the rendering implementation
export const WithClassName: Story = (args) => <Title {...args} />;
WithClassName.args = {
    children: 'Title with custom class',
    highlight: 'class',
    subtitle: 'Custom styling example',
    className: 'custom-title-class',
    align: "left",
    size: "xl",
    weight: "bold",
    color: "primary",
    as: "h1",
};