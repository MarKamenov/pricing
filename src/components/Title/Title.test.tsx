import { render, screen } from "@testing-library/react"
import { Title } from "../Title"

describe("Title component", () => {
    test('should render component', () => {
        const { container } = render(<Title>Hello World</Title>)

        expect(container).toBeInTheDocument();
    });

    test('should match snapshot', () => {
        const { container } = render(<Title>Hello World</Title>);

        expect(container).toMatchSnapshot();
    });

    describe("Basic Rendering", () => {
        test("renders basic title without highlights", () => {
            render(<Title>Hello World</Title>)
            expect(screen.getByText("Hello World")).toBeInTheDocument()
        })

        test("renders with correct default props", () => {
            render(<Title>Default Props</Title>)
            const title = screen.getByText("Default Props")
            expect(title.tagName).toBe("H1")
            expect(title).toHaveClass("title", "align-left", "size-xl", "weight-bold", "color-primary")
        })
    })

    describe("Highlight Functionality", () => {
        test("renders with single highlight", () => {
            render(<Title highlight="World">Hello World</Title>)
            expect(screen.getByText("World")).toHaveClass("highlight")
        })

        test("renders with multiple highlights", () => {
            render(<Title highlight={["Hello", "World"]}>Hello Beautiful World</Title>)
            expect(screen.getByText("Hello")).toHaveClass("highlight")
            expect(screen.getByText("World")).toHaveClass("highlight")
        })

        test("handles case-insensitive highlights", () => {
            render(<Title highlight="WORLD">Hello World</Title>)
            expect(screen.getByText("World")).toHaveClass("highlight")
        })

        test("handles empty highlight array", () => {
            render(<Title highlight={[]}>Hello World</Title>)
            expect(screen.getByText("Hello World")).toBeInTheDocument()
        })

        test("handles highlight not found in text", () => {
            render(<Title highlight="Foo">Hello World</Title>)
            expect(screen.getByText("Hello World")).toBeInTheDocument()
        })
    })

    describe("Style Variations", () => {
        test("applies different alignments", () => {
            const { rerender } = render(<Title align="center">Centered Title</Title>)
            expect(screen.getByText("Centered Title")).toHaveClass("align-center")

            rerender(<Title align="right">Right Title</Title>)
            expect(screen.getByText("Right Title")).toHaveClass("align-right")
        })

        test("applies different sizes", () => {
            const { rerender } = render(<Title size="sm">Small Title</Title>)
            expect(screen.getByText("Small Title")).toHaveClass("size-sm")

            rerender(<Title size="3xl">Large Title</Title>)
            expect(screen.getByText("Large Title")).toHaveClass("size-3xl")
        })

        test("applies different weights", () => {
            const { rerender } = render(<Title weight="normal">Normal Weight</Title>)
            expect(screen.getByText("Normal Weight")).toHaveClass("weight-normal")

            rerender(<Title weight="semibold">Semibold Weight</Title>)
            expect(screen.getByText("Semibold Weight")).toHaveClass("weight-semibold")
        })

        test("applies different colors", () => {
            const { rerender } = render(<Title color="accent">Accent Color</Title>)
            expect(screen.getByText("Accent Color")).toHaveClass("color-accent")

            rerender(<Title color="secondary">Secondary Color</Title>)
            expect(screen.getByText("Secondary Color")).toHaveClass("color-secondary")
        })
    })

    describe("Component Variations", () => {
        test("renders different heading levels", () => {
            const { rerender } = render(<Title as="h2">H2 Title</Title>)
            expect(screen.getByText("H2 Title").tagName).toBe("H2")

            rerender(<Title as="h6">H6 Title</Title>)
            expect(screen.getByText("H6 Title").tagName).toBe("H6")
        })

        test("applies custom className", () => {
            render(<Title className="custom-class">Custom Class Title</Title>)
            expect(screen.getByText("Custom Class Title")).toHaveClass("custom-class")
        })
    })

    describe("Edge Cases", () => {
        test("handles undefined children", () => {
            render(<Title>{undefined}</Title>)
            expect(screen.getByRole("heading")).toBeInTheDocument()
            expect(screen.getByRole("heading")).toHaveTextContent("")
        })

        test("handles null children", () => {
            render(<Title>{null}</Title>)
            expect(screen.getByRole("heading")).toBeInTheDocument()
            expect(screen.getByRole("heading")).toHaveTextContent("")
        })

        test("handles empty string children", () => {
            render(<Title>{""}</Title>)
            expect(screen.getByRole("heading")).toBeInTheDocument()
            expect(screen.getByRole("heading")).toHaveTextContent("")
        })

        test("handles number children", () => {
            render(<Title>{42}</Title>)
            expect(screen.getByText("42")).toBeInTheDocument()
        })
    })
})

