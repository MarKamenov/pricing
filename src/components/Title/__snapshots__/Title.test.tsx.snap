// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title component should match snapshot 1`] = `
<div>
  <h1
    class="title align-left size-xl weight-bold color-primary"
  >
    Hello World
  </h1>
</div>
`;
